In [ ]:
!git clone https://github.com/CMU-MultiComp-Lab/CMU-MultimodalSDK.git

Cloning into 'CMU-MultimodalSDK'...
remote: Enumerating objects: 100, done.
remote: Counting objects: 100% (100/100), done.
remote: Compressing objects: 100% (82/82), done.
remote: Total 100 (delta 13), reused 95 (delta 11), pack-reused 0 (from 0)
Receiving objects: 100% (100/100), 294.05 KiB | 2.58 MiB/s, done.
Resolving deltas: 100% (13/13), done.


In [ ]:
%cd /content/CMU-MultimodalSDK/
!pip install -e .

/content/CMU-MultimodalSDK
Obtaining file:///content/CMU-MultimodalSDK
  Preparing metadata (setup.py) ... done
  Attempting uninstall: mmsdk
    Found existing installation: mmsdk 1.1.0
    Uninstalling mmsdk-1.1.0:
      Successfully uninstalled mmsdk-1.1.0
  Running setup.py develop for mmsdk


In [ ]:
from mmsdk import mmdatasdk
DATA_PATH = "./cmu_mosei/"
cmu_mosei_highlevel = mmdatasdk.mmdataset(mmdatasdk.cmu_mosei.highlevel, DATA_PATH)

[2025-11-21 00:51:04.172] | Status  | Downloading from http://immortal.multicomp.cs.cmu.edu/CMU-MOSEI/language/CMU_MOSEI_TimestampedWordVectors.csd to ./cmu_mosei/CMU_MOSEI_TimestampedWordVectors.csd...


[2025-11-21 00:54:02.433] | Success | Download complete!
[2025-11-21 00:54:02.435] | Success | Computational sequence read from file ./cmu_mosei/CMU_MOSEI_TimestampedWordVectors.csd ...
[2025-11-21 00:54:02.547] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2025-11-21 00:54:02.547] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...


[2025-11-21 00:54:04.165] | Success | <glove_vectors> computational sequence data in correct format.
[2025-11-21 00:54:04.165] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2025-11-21 00:54:04.165] | Warning | <glove_vectors> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 00:54:04.888] | Status  | Downloading from http://immortal.multicomp.cs.cmu.edu/CMU-MOSEI/acoustic/CMU_MOSEI_COVAREP.csd to ./cmu_mosei/CMU_MOSEI_COVAREP.csd...


[2025-11-21 01:18:45.047] | Success | Download complete!
[2025-11-21 01:18:45.051] | Success | Computational sequence read from file ./cmu_mosei/CMU_MOSEI_COVAREP.csd ...
[2025-11-21 01:18:45.936] | Status  | Checking the integrity of the <COVAREP> computational sequence ...
[2025-11-21 01:18:45.936] | Status  | Checking the format of the data in <COVAREP> computational sequence ...


[2025-11-21 01:18:48.420] | Success | <COVAREP> computational sequence data in correct format.
[2025-11-21 01:18:48.420] | Status  | Checking the format of the metadata in <COVAREP> computational sequence ...
[2025-11-21 01:18:48.420] | Warning | <COVAREP> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:18:49.280] | Status  | Downloading from http://immortal.multicomp.cs.cmu.edu/CMU-MOSEI/visual/CMU_MOSEI_VisualOpenFace2.csd to ./cmu_mosei/CMU_MOSEI_VisualOpenFace2.csd...


[2025-11-21 01:52:05.726] | Success | Download complete!
[2025-11-21 01:52:05.733] | Success | Computational sequence read from file ./cmu_mosei/CMU_MOSEI_VisualOpenFace2.csd ...
[2025-11-21 01:52:07.920] | Status  | Checking the integrity of the <OpenFace_2> computational sequence ...
[2025-11-21 01:52:07.920] | Status  | Checking the format of the data in <OpenFace_2> computational sequence ...


[2025-11-21 01:52:11.997] | Success | <OpenFace_2> computational sequence data in correct format.
[2025-11-21 01:52:11.997] | Status  | Checking the format of the metadata in <OpenFace_2> computational sequence ...
[2025-11-21 01:52:11.997] | Warning | <OpenFace_2> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:52:12.746] | Status  | Downloading from http://immortal.multicomp.cs.cmu.edu/CMU-MOSEI/visual/CMU_MOSEI_VisualFacet42.csd to ./cmu_mosei/CMU_MOSEI_VisualFacet42.csd...


[2025-11-21 01:55:16.431] | Success | Download complete!
[2025-11-21 01:55:16.433] | Success | Computational sequence read from file ./cmu_mosei/CMU_MOSEI_VisualFacet42.csd ...
[2025-11-21 01:55:16.544] | Status  | Checking the integrity of the <FACET 4.2> computational sequence ...
[2025-11-21 01:55:16.544] | Status  | Checking the format of the data in <FACET 4.2> computational sequence ...


[2025-11-21 01:55:18.581] | Success | <FACET 4.2> computational sequence data in correct format.
[2025-11-21 01:55:18.581] | Status  | Checking the format of the metadata in <FACET 4.2> computational sequence ...
[2025-11-21 01:55:18.581] | Warning | <FACET 4.2> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:55:18.581] | Success | Dataset initialized successfully ... 


In [ ]:
cmu_mosei_labels = mmdatasdk.mmdataset(mmdatasdk.cmu_mosei.labels, DATA_PATH)

[2025-11-21 01:55:19.379] | Status  | Downloading from http://immortal.multicomp.cs.cmu.edu/CMU-MOSEI/labels/CMU_MOSEI_Labels.csd to ./cmu_mosei/CMU_MOSEI_Labels.csd...


[2025-11-21 01:55:22.952] | Success | Download complete!
[2025-11-21 01:55:22.953] | Success | Computational sequence read from file ./cmu_mosei/CMU_MOSEI_Labels.csd ...
[2025-11-21 01:55:23.104] | Status  | Checking the integrity of the <All Labels> computational sequence ...
[2025-11-21 01:55:23.104] | Status  | Checking the format of the data in <All Labels> computational sequence ...


[2025-11-21 01:55:24.892] | Success | <All Labels> computational sequence data in correct format.
[2025-11-21 01:55:24.892] | Status  | Checking the format of the metadata in <All Labels> computational sequence ...
[2025-11-21 01:55:24.892] | Warning | <All Labels> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:55:24.892] | Success | Dataset initialized successfully ... 


In [ ]:
%cd /content/CMU-MultimodalSDK/

/content/CMU-MultimodalSDK


In [ ]:
import argparse
import numpy as np
from os import listdir
from os.path import isfile, join
import os
from mmsdk import mmdatasdk

In [ ]:
DATA_PATH = "/content/cmu_mosei/"
parser = argparse.ArgumentParser(description='Reading dataset by files')
parser.add_argument('--cmu_mosei_path', type=str, default=DATA_PATH,
                    help='the folder path to read dataset from')
args = parser.parse_args(args=[]) # Pass an empty list to avoid parsing kernel arguments
dataset_dictionary = {}
if os.path.isdir(args.cmu_mosei_path) is False:
	print ("Folder does not exist ...")
	exit(-1)

csdfiles = [f for f in listdir(args.cmu_mosei_path) if isfile(join(args.cmu_mosei_path, f)) and f[-4:]=='.csd']
if len(csdfiles)==0:
	print("No csd files in the given folder")
	exit(-2)

print("%d csd files found"%len(csdfiles))
for csdfile in csdfiles:
	dataset_dictionary[csdfile]=os.path.join(args.cmu_mosei_path,csdfile)
dataset=mmdatasdk.mmdataset(dataset_dictionary)



print ("List of the computational sequences")
print (dataset.computational_sequences.keys())

5 csd files found
[2025-11-21 01:57:57.199] | Success | Computational sequence read from file /content/cmu_mosei/CMU_MOSEI_VisualFacet42.csd ...
[2025-11-21 01:57:57.503] | Status  | Checking the integrity of the <FACET 4.2> computational sequence ...
[2025-11-21 01:57:57.503] | Status  | Checking the format of the data in <FACET 4.2> computational sequence ...


[2025-11-21 01:57:59.166] | Success | <FACET 4.2> computational sequence data in correct format.
[2025-11-21 01:57:59.166] | Status  | Checking the format of the metadata in <FACET 4.2> computational sequence ...
[2025-11-21 01:57:59.166] | Warning | <FACET 4.2> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:57:59.167] | Success | Computational sequence read from file /content/cmu_mosei/CMU_MOSEI_Labels.csd ...
[2025-11-21 01:57:59.241] | Status  | Checking the integrity of the <All Labels> computational sequence ...
[2025-11-21 01:57:59.241] | Status  | Checking the format of the data in <All Labels> computational sequence ...


[2025-11-21 01:58:00.672] | Success | <All Labels> computational sequence data in correct format.
[2025-11-21 01:58:00.672] | Status  | Checking the format of the metadata in <All Labels> computational sequence ...
[2025-11-21 01:58:00.672] | Warning | <All Labels> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:58:00.673] | Success | Computational sequence read from file /content/cmu_mosei/CMU_MOSEI_COVAREP.csd ...
[2025-11-21 01:58:00.973] | Status  | Checking the integrity of the <COVAREP> computational sequence ...
[2025-11-21 01:58:00.974] | Status  | Checking the format of the data in <COVAREP> computational sequence ...


[2025-11-21 01:58:04.632] | Success | <COVAREP> computational sequence data in correct format.
[2025-11-21 01:58:04.633] | Status  | Checking the format of the metadata in <COVAREP> computational sequence ...
[2025-11-21 01:58:04.633] | Warning | <COVAREP> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:58:04.634] | Success | Computational sequence read from file /content/cmu_mosei/CMU_MOSEI_TimestampedWordVectors.csd ...
[2025-11-21 01:58:04.894] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2025-11-21 01:58:04.894] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...


[2025-11-21 01:58:10.905] | Success | <glove_vectors> computational sequence data in correct format.
[2025-11-21 01:58:10.906] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2025-11-21 01:58:10.906] | Warning | <glove_vectors> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:58:10.906] | Success | Computational sequence read from file /content/cmu_mosei/CMU_MOSEI_VisualOpenFace2.csd ...
[2025-11-21 01:58:11.003] | Status  | Checking the integrity of the <OpenFace_2> computational sequence ...
[2025-11-21 01:58:11.003] | Status  | Checking the format of the data in <OpenFace_2> computational sequence ...


[2025-11-21 01:58:12.569] | Success | <OpenFace_2> computational sequence data in correct format.
[2025-11-21 01:58:12.569] | Status  | Checking the format of the metadata in <OpenFace_2> computational sequence ...
[2025-11-21 01:58:12.569] | Warning | <OpenFace_2> computational sequence does not have all the required metadata ... continuing 
[2025-11-21 01:58:12.569] | Success | Dataset initialized successfully ... 
List of the computational sequences
dict_keys(['CMU_MOSEI_VisualFacet42.csd', 'CMU_MOSEI_Labels.csd', 'CMU_MOSEI_COVAREP.csd', 'CMU_MOSEI_TimestampedWordVectors.csd', 'CMU_MOSEI_VisualOpenFace2.csd'])


In [ ]:
# After download data, populate and allign
labels = dataset["CMU_MOSEI_Labels.csd"]
visuals = dataset["CMU_MOSEI_VisualFacet42.csd"]
audios = dataset["CMU_MOSEI_COVAREP.csd"]
texts = dataset["CMU_MOSEI_TimestampedWordVectors.csd"]

In [ ]:
list(labels.data.keys())[0]

'--qXJuDtHPw'

In [ ]:
labels["--qXJuDtHPw"]

<HDF5 group "/All Labels/data/--qXJuDtHPw" (2 members)>

In [ ]:
# It is recommended to allign everything by text. This is the most common approach
# and also recommended as mmsdk is optimized for it
## BEFORE allign
sample_key1 = list(labels.data.keys())[10]
visual_seg = list(visuals.data[sample_key1].keys())
audios_seg = list(audios.data[sample_key1].keys())

In [ ]:
print(np.array(visuals.data[sample_key1]["intervals"]))
print(np.array(audios.data[sample_key1]["intervals"]))

[[0.00000e+00 3.33333e-02]
 [3.33333e-02 6.66667e-02]
 [6.66667e-02 1.00000e-01]
 ...
 [1.01467e+02 1.01500e+02]
 [1.01500e+02 1.01533e+02]
 [1.01533e+02 1.01567e+02]]
[[0.0000e+00 1.0000e-02]
 [1.0000e-02 2.0000e-02]
 [2.0000e-02 3.0000e-02]
 ...
 [1.0160e+02 1.0161e+02]
 [1.0161e+02 1.0162e+02]
 [1.0162e+02 1.0163e+02]]


As we can see that the visuals has different interval. Sample 1 has interval of $3.33*10^{-2}$ whereas sample 2 has interval of $4.0*10^{-2}$

In [ ]:
sample_key2 = list(labels.data.keys())[19]
print(np.array(visuals.data[sample_key2]["intervals"]))
print(np.array(audios.data[sample_key2]["intervals"]))


[[0.0000e+00 4.0000e-02]
 [4.0000e-02 8.0000e-02]
 [8.0000e-02 1.2000e-01]
 ...
 [5.6984e+02 5.6988e+02]
 [5.6988e+02 5.6992e+02]
 [5.6992e+02 5.6996e+02]]
[[0.0000e+00 1.0000e-02]
 [1.0000e-02 2.0000e-02]
 [2.0000e-02 3.0000e-02]
 ...
 [5.6997e+02 5.6998e+02]
 [5.6998e+02 5.6999e+02]
 [5.6999e+02 5.7000e+02]]


Since the Dataset is too large, we would want to truncate them first before aligning those visuals and text. We will do it using labels then use the key from label to extract for what we need. While extracting, we try to balanced the number of emotions.

In [ ]:
import math

In [ ]:
# 'dimension names': ['sentiment','happy','sad','anger','surprise','disgust','fear'],
emotions = ['happy','sad','anger','surprise','disgust','fear']
def is_equal(a, b):
  return b >= 0.1 and abs(a - b) < 1e-5
def extract_emotion_sentiment(arr):
  np_arr = np.array(arr)[1:]
  indices = np.where(is_equal(np_arr, np_arr.max()))
  indication_vector = np.zeros(6, dtype=int)
  indication_vector[indices] = 1
  return indication_vector

In [ ]:
# function to get the emotion and sentiment
def get_emotion(label_key):
  label = labels.data[label_key]
  features = np.array(label["features"])
  intervals = np.array(label["intervals"])
  emotions = np.zeros(6, dtype=int)
  for i in range(min(len(features) , len(intervals))):
    emotions += extract_emotion_sentiment(features[i])
  return emotions

In [ ]:
key_list = list(labels.data.keys())
emotion_sum = []
for x in key_list:
  emotion_sum.append(get_emotion(x))
print(np.array(emotion_sum))

/tmp/ipython-input-4101884438.py:7: DeprecationWarning: Calling nonzero on 0d arrays is deprecated, as it behaves surprisingly. Use `atleast_1d(cond).nonzero()` if the old behavior was intended. If the context of this warning is of the form `arr[nonzero(cond)]`, just use `arr[cond]`.
  indices = np.where(is_equal(np_arr, np_arr.max()))


[[1 0 0 0 0 0]
 [4 4 0 0 0 1]
 [3 0 0 0 0 0]
 ...
 [5 0 0 0 0 0]
 [6 0 0 0 0 0]
 [1 0 0 0 0 0]]


In [ ]:
col_sum = np.sum(emotion_sum, axis=0)
print(col_sum)

[11081  4307  3491   926  2402   987]


In [ ]:
balance_emotion = []
balance_key = []
for x in range(len(emotion_sum)):
  if (emotion_sum[x][-3] > 0 or emotion_sum[x][-1] > 0):
    balance_emotion.append(emotion_sum[x])
    balance_key.append(key_list[x])

In [ ]:
print(np.sum(balance_emotion, axis=0))
print(balance_key)
print(len(balance_key))

[4565 2210 1671  926 1320  987]
['-3g5yACwYnA', '-9y-fZ3swSY', '-AUZQgSxyPQ', '-Alixo7euuU', '-HvKLjmsO5U', '-HwX2H8Z4hY', '-I_e4mIh0yE', '-IqSFQePnpU', '-RpZEe4w4fY', '-UUCSKoHeMA', '-WXXTNIJcVM', '-ZgjBOA1Yhw', '-aNfi7CP8vM', '-aqamKhZ1Ec', '-dxfTGcXJoc', '-hnBHBN8p5A', '-lzEya4AM_4', '-qMhGeJkp5o', '-tPCytz4rww', '03X1FwF6udc', '06st0Ez_E_I', '0BVed2nBq1g', '0EAAfLCQabY', '0K7dCp80n9c', '0OC3wf9x3zQ', '0OXeN3MhFeA', '0SfejBuLgFo', '0YdYFtVdlWE', '0cBmQ_ItVUc', '0jtdrsmVzYQ', '0p9XKC_J3hU', '0tkWa2skSmA', '0y022OlZ3W0', '0zl_GDVzykw', '101787', '101851', '102184', '10219', '102213', '102408', '104741', '106077', '106514', '106941', '107585', '108793', '112425', '112433', '112509', '11650', '119348', '121358', '121400', '122439', '122602', '125676', '125845', '128258', '128752', '130149', '130366', '130426', '130448', '130456', '130633', '131650', '131871', '134252', '135623', '136196', '136205', '136211', '136215', '139006', '140293', '154449', '15Wu4U2VPLw', '167521', '173CpFb3clw',

Now that we have truncated our dataset. Let write it to a new directory for preservation.

In [ ]:
from mmsdk.mmdatasdk import computational_sequence

for key in key_list:
  if key not in balance_key:
    try:
      del labels.data[key]
      del visuals.data[key]
      del audios.data[key]
      del texts.data[key]
    except:
      continue

In [ ]:
len(labels.data.keys())

1186

In [ ]:
NEW_DATAPATH = "/content/filter_mosei/"
features_path = {
    "audio":  NEW_DATAPATH + "COVAREP.csd",
    "visual": NEW_DATAPATH + "VisualFacet42.csd",
    "text":   NEW_DATAPATH + "TimestampedWordVectors.csd"
}

labels_path = {
    "labels": NEW_DATAPATH + "Labels.csd"
}


In [ ]:
labels.deploy(labels_path["labels"])
visuals.deploy(features_path["visual"])
audios.deploy(features_path["audio"])
texts.deploy(features_path["text"])

[2025-11-21 05:40:56.426] | Status  | Checking the integrity of the <All Labels> computational sequence ...
[2025-11-21 05:40:56.426] | Status  | Checking the format of the data in <All Labels> computational sequence ...


[2025-11-21 05:40:56.930] | Success | <All Labels> computational sequence data in correct format.
[2025-11-21 05:40:56.930] | Status  | Checking the format of the metadata in <All Labels> computational sequence ...
[2025-11-21 05:40:56.930] | Success | <All Labels> computational sequence metadata in correct format.
[2025-11-21 05:40:56.930] | Success | <All Labels> computational sequence is valid!
[2025-11-21 05:40:56.930] | Status  | Deploying the </content/filter_mosei/Labels.csd> computational sequence to All Labels
[2025-11-21 05:40:56.930] | Status  | Your unique identifier for <All Labels> computational sequence is bfad70f4-8357-40ff-86e9-5d3f086e0a5e
[2025-11-21 05:40:56.930] | Status  | Writing the <All Labels> computational sequence data to /content/filter_mosei/Labels.csd
[2025-11-21 05:40:56.930] | Advise  | Compression with gzip and opts -9


[2025-11-21 05:40:58.566] | Success | <All Labels> computational sequence data successfully wrote to /content/filter_mosei/Labels.csd
[2025-11-21 05:40:58.566] | Status  | Writing the <All Labels> computational sequence metadata to /content/filter_mosei/Labels.csd
[2025-11-21 05:40:58.849] | Success | <All Labels> computational sequence metadata successfully wrote to /content/filter_mosei/Labels.csd
[2025-11-21 05:40:58.849] | Success | <All Labels> computational sequence successfully wrote to /content/filter_mosei/Labels.csd ...
[2025-11-21 05:40:58.850] | Input   | Please input dimension namescomputational sequence version for <FACET 4.2> computational sequence: filtered
[2025-11-21 05:41:07.213] | Status  | Checking the integrity of the <FACET 4.2> computational sequence ...
[2025-11-21 05:41:07.213] | Status  | Checking the format of the data in <FACET 4.2> computational sequence ...


[2025-11-21 05:41:08.374] | Success | <FACET 4.2> computational sequence data in correct format.
[2025-11-21 05:41:08.374] | Status  | Checking the format of the metadata in <FACET 4.2> computational sequence ...
[2025-11-21 05:41:08.375] | Success | <FACET 4.2> computational sequence metadata in correct format.
[2025-11-21 05:41:08.375] | Success | <FACET 4.2> computational sequence is valid!
[2025-11-21 05:41:08.375] | Status  | Deploying the </content/filter_mosei/VisualFacet42.csd> computational sequence to FACET 4.2
[2025-11-21 05:41:08.375] | Status  | Your unique identifier for <FACET 4.2> computational sequence is 448afa2c-d0d7-456c-89db-4e6dfc02d322
[2025-11-21 05:41:08.375] | Status  | Writing the <FACET 4.2> computational sequence data to /content/filter_mosei/VisualFacet42.csd
[2025-11-21 05:41:08.375] | Advise  | Compression with gzip and opts -9


[2025-11-21 05:41:56.529] | Success | <FACET 4.2> computational sequence data successfully wrote to /content/filter_mosei/VisualFacet42.csd
[2025-11-21 05:41:56.529] | Status  | Writing the <FACET 4.2> computational sequence metadata to /content/filter_mosei/VisualFacet42.csd
[2025-11-21 05:41:56.602] | Success | <FACET 4.2> computational sequence metadata successfully wrote to /content/filter_mosei/VisualFacet42.csd
[2025-11-21 05:41:56.602] | Success | <FACET 4.2> computational sequence successfully wrote to /content/filter_mosei/VisualFacet42.csd ...
[2025-11-21 05:41:56.602] | Input   | Please input dimension namescomputational sequence version for <COVAREP> computational sequence: filtered
[2025-11-21 05:43:37.880] | Status  | Checking the integrity of the <COVAREP> computational sequence ...
[2025-11-21 05:43:37.880] | Status  | Checking the format of the data in <COVAREP> computational sequence ...


[2025-11-21 05:43:39.062] | Success | <COVAREP> computational sequence data in correct format.
[2025-11-21 05:43:39.062] | Status  | Checking the format of the metadata in <COVAREP> computational sequence ...
[2025-11-21 05:43:39.062] | Success | <COVAREP> computational sequence metadata in correct format.
[2025-11-21 05:43:39.062] | Success | <COVAREP> computational sequence is valid!
[2025-11-21 05:43:39.062] | Status  | Deploying the </content/filter_mosei/COVAREP.csd> computational sequence to COVAREP
[2025-11-21 05:43:39.062] | Status  | Your unique identifier for <COVAREP> computational sequence is 80f2fddc-4a90-4fcb-9158-1e1939527808
[2025-11-21 05:43:39.062] | Status  | Writing the <COVAREP> computational sequence data to /content/filter_mosei/COVAREP.csd
[2025-11-21 05:43:39.062] | Advise  | Compression with gzip and opts -9


[2025-11-21 05:53:34.112] | Success | <COVAREP> computational sequence data successfully wrote to /content/filter_mosei/COVAREP.csd
[2025-11-21 05:53:34.113] | Status  | Writing the <COVAREP> computational sequence metadata to /content/filter_mosei/COVAREP.csd
[2025-11-21 05:53:34.383] | Success | <COVAREP> computational sequence metadata successfully wrote to /content/filter_mosei/COVAREP.csd
[2025-11-21 05:53:34.383] | Success | <COVAREP> computational sequence successfully wrote to /content/filter_mosei/COVAREP.csd ...
[2025-11-21 05:53:34.384] | Input   | Please input dimension namescomputational sequence version for <glove_vectors> computational sequence: filtered
[2025-11-21 05:53:54.111] | Status  | Checking the integrity of the <glove_vectors> computational sequence ...
[2025-11-21 05:53:54.112] | Status  | Checking the format of the data in <glove_vectors> computational sequence ...


[2025-11-21 05:53:55.274] | Success | <glove_vectors> computational sequence data in correct format.
[2025-11-21 05:53:55.274] | Status  | Checking the format of the metadata in <glove_vectors> computational sequence ...
[2025-11-21 05:53:55.274] | Success | <glove_vectors> computational sequence metadata in correct format.
[2025-11-21 05:53:55.274] | Success | <glove_vectors> computational sequence is valid!
[2025-11-21 05:53:55.274] | Status  | Deploying the </content/filter_mosei/TimestampedWordVectors.csd> computational sequence to glove_vectors
[2025-11-21 05:53:55.274] | Status  | Your unique identifier for <glove_vectors> computational sequence is febb6023-eb08-49f4-aa8d-c1e36a2631ce
[2025-11-21 05:53:55.275] | Status  | Writing the <glove_vectors> computational sequence data to /content/filter_mosei/TimestampedWordVectors.csd
[2025-11-21 05:53:55.275] | Advise  | Compression with gzip and opts -9


[2025-11-21 05:54:43.779] | Success | <glove_vectors> computational sequence data successfully wrote to /content/filter_mosei/TimestampedWordVectors.csd
[2025-11-21 05:54:43.779] | Status  | Writing the <glove_vectors> computational sequence metadata to /content/filter_mosei/TimestampedWordVectors.csd
[2025-11-21 05:54:43.872] | Success | <glove_vectors> computational sequence metadata successfully wrote to /content/filter_mosei/TimestampedWordVectors.csd
[2025-11-21 05:54:43.873] | Success | <glove_vectors> computational sequence successfully wrote to /content/filter_mosei/TimestampedWordVectors.csd ...


In [ ]:
token = "nPeroUpmd8M6BA2OTH0GZu7Br00xbn6aPT4bOYt0fIqoZmY2FlPo61WdN5VF"
BASE_URL = "https://zenodo.org/api"


In [ ]:
# ============================================
# COMPLETE UPLOAD SCRIPT - PASTE THIS IN COLAB
# ============================================

!pip install requests -q

import requests
import os
import json
from getpass import getpass

# Get your Zenodo token
# Corrected: Use a proper prompt for getpass, and ensure it gets a valid token.
# Alternatively, for Colab, you can use google.colab.userdata.get('ZENODO_TOKEN')
# or simply assign your token directly for testing purposes.
ZENODO_TOKEN = getpass("Enter your Zenodo API token: ")

BASE_URL = "https://zenodo.org/api"

def upload_mosei_to_zenodo(directory, token):
    """Upload all .csd files from directory to Zenodo"""

    print("="*70)
    print("UPLOADING TO ZENODO")
    print("="*70)

    # Create deposit
    print("\n[1/4] Creating deposit...")
    r = requests.post(f"{BASE_URL}/deposit/depositions",
                     params={"access_token": token}, json={})
    deposit = r.json()
    # Check for errors in the response before trying to access keys
    if r.status_code != 201: # 201 Created is the expected status for a new deposit
        print(f"API Error: {r.status_code} - {r.text}")
        # Attempt to get error details if available
        try:
            error_details = r.json()
            print(f"Error details: {error_details}")
        except json.JSONDecodeError:
            pass
        raise ValueError("Failed to create Zenodo deposit. Check your token and API call.")

    bucket = deposit["links"]["bucket"]
    deposit_id = deposit["id"]
    print(f"✓ Deposit ID: {deposit_id}")

    # Find and upload all .csd files
    print("\n[2/4] Uploading files...")
    csd_files = [f for f in os.listdir(directory) if f.endswith('.csd')]

    uploaded = []
    for csd_file in csd_files:
        filepath = os.path.join(directory, csd_file)
        size_gb = os.path.getsize(filepath) / (1024**3)
        print(f"  Uploading {csd_file} ({size_gb:.2f} GB)...", end=" ", flush=True)

        with open(filepath, "rb") as f:
            r = requests.put(f"{bucket}/{csd_file}",
                           data=f, params={"access_token": token})

        if r.status_code in [200, 201]:
            print("✓")
            uploaded.append(r.json())
        else:
            print(f"FAILED: {r.text}")
            raise ValueError(f"Failed to upload file {csd_file}: {r.text}")

    print(f"\n✓ Uploaded {len(uploaded)} files")

    # Add metadata
    print("\n[3/4] Adding metadata...")
    file_list = '\n'.join(f'- {f["key"]}' for f in uploaded)

    metadata = {
        "metadata": {
            "title": "MOSEI Processed Dataset - Multimodal Features",
            "upload_type": "dataset",
            "description": f"""Processed MOSEI dataset with multiple modalities.

**Files:**
{file_list}

**Usage:**
```python
from mmsdk import mmdatasdk
recipe = {{'Labels': 'DOWNLOAD_URL'}}
dataset = mmdatasdk.mmdataset(recipe)
```

Original: CMU-MOSEI (Zadeh et al., 2018)
SDK: CMU-MultimodalSDK
""",
            "creators": [{"name": "Your Name"}],
            "keywords": ["multimodal", "sentiment", "MOSEI", "audio", "visual", "text"],
            "access_right": "open",
            "license": "cc-by-4.0"
        }
    }

    r = requests.put(f"{BASE_URL}/deposit/depositions/{deposit_id}",
                    params={"access_token": token}, json=metadata)
    if r.status_code != 200: # 200 OK is expected for metadata update
        print(f"API Error: {r.status_code} - {r.text}")
        raise ValueError("Failed to add metadata to Zenodo deposit.")
    print("✓ Metadata added")

    # Publish
    print("\n[4/4] Publishing...")
    r = requests.post(f"{BASE_URL}/deposit/depositions/{deposit_id}/actions/publish",
                     params={"access_token": token})
    if r.status_code != 202: # 202 Accepted is expected for publish action
        print(f"API Error: {r.status_code} - {r.text}")
        raise ValueError("Failed to publish Zenodo deposit.")
    result = r.json()

    print("\n" + "="*70)
    print("✓✓✓ SUCCESS ✓✓✓")
    print("="*70)
    print(f"\nDOI: {result['doi']}")
    print(f"URL: {result['links']['record_html']}")

    print("\n📥 Download URLs:")
    print("-"*70)

# Run the upload
result = upload_mosei_to_zenodo("/content/filter_mosei", ZENODO_TOKEN)


UPLOADING TO ZENODO

[1/4] Creating deposit...
✓ Deposit ID: 17668236

[2/4] Uploading files...
  Uploading VisualFacet42.csd (0.73 GB)... ✓
  Uploading Labels.csd (0.01 GB)... ✓
  Uploading COVAREP.csd (4.99 GB)... ✓
  Uploading TimestampedWordVectors.csd (0.73 GB)... ✓

✓ Uploaded 4 files

[3/4] Adding metadata...
✓ Metadata added

[4/4] Publishing...

✓✓✓ SUCCESS ✓✓✓

DOI: 10.5281/zenodo.17668236
URL: https://zenodo.org/record/17668236

📥 Download URLs:
----------------------------------------------------------------------


KeyError: 'key'

NameError: name 'result' is not defined